gax/operation.h

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GAPIC_GENERATOR_CPP_GAX_OPERATION_H_
+#define GAPIC_GENERATOR_CPP_GAX_OPERATION_H_
+
+#include "call_context.h"
+#include "internal/gtest_prod.h"
+#include "operations_stub.h"
+#include "status.h"
+#include "status_or.h"
+
+#include "google/longrunning/operations.pb.h"
+
+#include <iostream>
+
+#include <memory>
+
+namespace google {
+
+namespace gax {
+
+// Note: compilation will fail unless ResponseT and MetadataT are children of
+// protobuf::Message.
+template <typename ResponseT, typename MetadataT>
+class Operation final {
+ public:
+  Operation(std::shared_ptr<gax::OperationsStub> stub,
+            google::longrunning::Operation op)
+      : stub_(std::move(stub)), op_(std::move(op)) {}
+  Operation() = delete;
+  Operation(Operation&& rhs) = default;
+
+  std::string const& Name() { return op_.name(); }
+
+  gax::StatusOr<ResponseT> Result() {
+    if (!Done()) {
+      return gax::Status{gax::StatusCode::kUnknown,
+                         "operation has not completed=" + Name()};
+    } else if (op_.has_error()) {
+      return gax::Status{static_cast<gax::StatusCode>(op_.error().code()),
+                         op_.error().message()};
+    } else {
+      auto const& any = op_.response();
+      if (!any.Is<ResponseT>()) {
+        return gax::Status{gax::StatusCode::kUnknown,
+                           "invalid result in operation=" + Name()};
+      }
+
+      ResponseT result;
+      any.UnpackTo(&result);
+      return std::move(result);
+    }
+  }
+
+  MetadataT Metadata() const {
+    MetadataT m;
+    op_.metadata().UnpackTo(&m);
+    return m;
+  }
+
+  gax::Status Update() {
+    if (!Done()) {
+      google::longrunning::GetOperationRequest request;
+      request.set_name(op_.name());
+      gax::CallContext context(OperationsStub::get_operation_info_);
+      return stub_->GetOperation(context, request, &op_);
+    } else if (op_.has_error()) {
+      return gax::Status{static_cast<gax::StatusCode>(op_.error().code()),
```

Comment:
Consider writing a `gax::Status::FromGrpc()` function or something similar, because you will write code like this in a lot of places.  Also, the static cast for error codes is suspect.

Reply:
At that point it might as well be an explicit constructor. I'll move it into the class.
The error codes for gax::StatusCode are explicitly and deliberately chosen to match those from gRPC. I'd hoped to save a few lines with the static cast, but if you think explicit translation is better I'll turn it into a switch.

Reply:
Added `GrpcStatusToGaxStatus` in #34 , probably going to rename that. Will update usage here depending on which PR lands first.